core/src/main/scala/quasar/datasource/kafka/TunnelSession.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2020 Precog Data
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.datasource.kafka
+
+import slamdata.Predef._
+
+import java.util.concurrent.atomic.AtomicReference
+
+import com.jcraft.jsch.Session
+
+final class TunnelSession private (session: Session) {
+  private[this] val tunnelsRef: AtomicReference[List[((String, Int), Int)]] = new AtomicReference(Nil)
+  def ports: List[Int] = tunnelsRef.get().map(_._2)
+  def resolve(host: String, port: Int): Int = {
```

Comment:
I haven't fully figured out what this function is doing yet. @dcsobral could you explain what this function is doing in a couple sentences? What situations is it steeling against by using the `while` loop in this way?

Reply:
I've refactored it in a way that should make what is happening clear, plus adding scaladocs about it. I wasn't happy with the implementation, but at the time I had a mental block on alternatives.